src/cpu/flexcpu/simple_dataflow_cpu.hh

```diff
@@ -0,0 +1,496 @@
+        {
+            callback(fault, req);
+
+            // It requires assumptions that users of this will be well-behaved
+            // to call "delete this;" in a C++ class, but since this is an
+            // internal class, we should be able to contain access to this line
+            if (selfDestruct) delete this;
+        }
+    };
+
+    // END Internal class definitions
+
+    // BEGIN Internal container structs
+
+    struct TranslationReq
+    {
+        RequestPtr request;
+        TranslationCallback callback;
+    };
+
+    struct FetchReq
+    {
+        PacketPtr packet;
+        FetchCallback callback;
+    };
+
+    struct ExecutionReq
+    {
+        // Note: Storing references since it is not this class' responsibility
+        //       to manage lifetime of the object memory used here.
+        StaticInstPtr staticInst;
+        std::weak_ptr<ExecContext> execContext;
+        ExecCallback callback;
+    };
+
+    struct SplitAccCtrlBlk
+    {
+        PacketPtr main = nullptr;
+        PacketPtr low = nullptr;
+        bool lowReceived = false;
+        PacketPtr high = nullptr;
+        bool highReceived = false;
+    };
+
+    struct MemAccessReq
+    {
+        PacketPtr packet;
+        StaticInstPtr staticInst;
+        std::weak_ptr<ExecContext> execContext;
+        ThreadContext* tc;
+        MemCallback callback;
+
+        SplitAccCtrlBlk* split;
+    };
+
+    // END Internal container structs
+
+
+    // BEGIN Internal constants
+
+    const Addr cacheBlockMask;
+
+    // END Internal constants
+
+
+    // BEGIN Internal state variables
+
+    DataPort _dataPort;
+    InstPort _instPort;
+
+    // Using unique_ptr for now since holding objects directly results
+    // in calls to the object's copy constructor, which SimpleThread does not
+    // like.
+    std::vector<std::unique_ptr<SDCPUThread>> threads;
+
+    // For use in enqueuing requests if we want to limit number of requests
+    // servicable at once in the future. TODO for now.
+    std::list<TranslationReq> dataTranslationReqs;
+    std::list<TranslationReq> instTranslationReqs;
+    std::list<FetchReq> fetchReqs;
+    std::list<ExecutionReq> executionReqs;
+    std::list<MemAccessReq> memReqs;
+
+    // Note: I was informed that the pointer you get back is the same pointer
+    //       you send through the port, so it's safe to do this mapping. It may
+    //       be safer to do this mapping through the SenderState struct instead
+    //       based on what has been learned since the writing of this though.
+    std::unordered_map<PacketPtr, FetchCallback> outstandingFetches;
+    std::unordered_map<PacketPtr, MemAccessReq> outstandingMemReqs;
+
+    // TODO temporary mark all executions as taking one cycle's length at least
+    Tick executionTime = clockPeriod();
+
+    // END Internal state variables
+
+
+    // BEGIN Internal functions
+
+    void attemptAllFetchReqs();
+
+    void attemptAllMemReqs();
```

Comment:
Well it's limited right now by the amount the port can accept at once.

I like the "attempt" terminology here since my intention is to simply try if possible to send them all through the port. It may also send none if the port is busy, so I don't really think a hard sendX scheme makes sense here.